x/leverage/keeper/keeper.go

```diff
@@ -270,3 +270,166 @@ func (k Keeper) GetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddres
 	key := types.CreateCollateralSettingKey(borrowerAddr, denom)
 	return store.Has(key)
 }
+
+// LiquidateBorrow attempts to repay one of an eligible borrower's borrows (in part or in full) in exchange
+// for a selected denomination of uToken collateral. If the borrower is not over their borrow limit, or
+// the repayment or reward denominations are invalid, an error is returned. If the attempted repayment
+// is greater than the amount owed or the maximum that can be repaid due to parameters (close factor)
+// then a partial liquidation, equal to the maximum valid amount, is performed. The same occurs if the
+// value of collateral in the selected reward denomination cannot cover the proposed repayment.
+func (k Keeper) LiquidateBorrow(
+	ctx sdk.Context, liquidatorAddr, borrowerAddr sdk.AccAddress, repayment sdk.Coin, rewardDenom string,
+) error {
+	if !repayment.IsValid() {
+		return sdkerrors.Wrap(types.ErrInvalidAsset, repayment.String())
+	}
+	if !k.IsAcceptedUToken(ctx, rewardDenom) {
+		return sdkerrors.Wrap(types.ErrInvalidAsset, rewardDenom)
+	}
+
+	// Get total borrowed by borrower (all denoms)
+	borrowed, err := k.GetBorrowerBorrows(ctx, borrowerAddr)
+	if err != nil {
+		return err
+	}
+
+	// Get borrower uToken balances, for all uToken denoms enabled as collateral
+	collateral := k.GetBorrowerCollateral(ctx, borrowerAddr)
+	if err != nil {
+		return err
+	}
+
+	// Use oracle helper functions to find total borrowed value in USD
+	borrowValue, err := k.GetTotalPrice(ctx, borrowed)
+	if err != nil {
+		return err
+	}
+
+	// Use collateral weights to compute borrow limit from enabled collateral
+	borrowLimit, err := k.CalculateBorrowLimit(ctx, collateral)
+	if err != nil {
+		return err
+	}
+
+	// Confirm borrower's eligibility for liquidation
+	if borrowLimit.GTE(borrowValue) {
+		return sdkerrors.Wrap(types.ErrLiquidationIneligible, borrowerAddr.String())
+	}
+
+	// Get reward-specific incentive and dynamic close factor
+	baseRewardDenom := k.FromUTokenToTokenDenom(ctx, rewardDenom)
+	liquidationIncentive, closeFactor, err := k.LiquidationParams(ctx, baseRewardDenom, borrowValue, borrowLimit)
+	if err != nil {
+		return err
+	}
+
+	// Repayment cannot exceed liquidator's available balance
+	liquidatorBalance := k.bankKeeper.GetBalance(ctx, liquidatorAddr, repayment.Denom)
+	if repayment.Amount.GTE(liquidatorBalance.Amount) {
+		repayment.Amount = liquidatorBalance.Amount
+	}
+
+	// Repayment cannot exceed borrower's borrowed amount of selected denom
+	if repayment.Amount.GTE(borrowed.AmountOf(repayment.Denom)) {
+		repayment.Amount = borrowed.AmountOf(repayment.Denom)
+	}
+
+	// Repayment cannot exceed borrowed value * close factor
+	repayValue, err := k.GetPrice(ctx, repayment)
+	if err != nil {
+		return err
+	}
+	if repayValue.GTE(borrowValue.Mul(closeFactor)) {
```

Comment:
We call `borrowValue.Mul(closeFactor)` twice. Let's put `maxRepayValue := borrowValue.Mul(closeFactor)` outside this `if` statement and use that in both places.